mmv1/third_party/terraform/services/storage/resource_storage_folder.go.tmpl

```diff
@@ -0,0 +1,358 @@
+	var listError, deleteObjectError error
+	for deleteObjectError == nil {
+		res, err := config.NewStorageClient(userAgent).Objects.List(bucket).Prefix(folderName[0] + "/").Do()
+		if err != nil {
+			log.Printf("Error listing contents of bucket %s: %v", bucket, err)
+			// If we can't list the contents, try deleting the bucket anyway in case it's empty
+			listError = err
+			break
+		}
+
+		if len(res.Items) == 0 {
+			break // 0 items, folder empty
+		}
+
+		if !d.Get("force_destroy").(bool) {
+			deleteErr := fmt.Errorf("Error trying to delete folder %s containing objects without `force_destroy` set to true", bucket)
+			log.Printf("Error! %s : %s\n\n", bucket, deleteErr)
+			return deleteErr
+		}
+		// GCS requires that a bucket be empty (have no objects or object
+		// versions) before it can be deleted.
+		log.Printf("[DEBUG] GCS Folder attempting to forceDestroy\n\n")
+
+		// Create a workerpool for parallel deletion of resources. In the
+		// future, it would be great to expose Terraform's global parallelism
+		// flag here, but that's currently reserved for core use. Testing
+		// shows that NumCPUs-1 is the most performant on average networks.
+		//
+		// The challenge with making this user-configurable is that the
+		// configuration would reside in the Terraform configuration file,
+		// decreasing its portability. Ideally we'd want this to connect to
+		// Terraform's top-level -parallelism flag, but that's not plumbed nor
+		// is it scheduled to be plumbed to individual providers.
+		wp := workerpool.New(runtime.NumCPU() - 1)
+
+		for _, object := range res.Items {
+			log.Printf("[DEBUG] Found %s", object.Name)
+			object := object
+
+			wp.Submit(func() {
+				log.Printf("[TRACE] Attempting to delete %s", object.Name)
+				if err := config.NewStorageClient(userAgent).Objects.Delete(bucket, object.Name).Generation(object.Generation).Do(); err != nil {
+					deleteObjectError = err
+					log.Printf("[ERR] Failed to delete storage object %s: %s", object.Name, err)
+				} else {
+					log.Printf("[TRACE] Successfully deleted %s", object.Name)
+				}
+			})
+		}
+
+		// Wait for everything to finish.
+		wp.StopWait()
+	}
+
+	if listError != nil {
+		return fmt.Errorf("could not delete non-empty folder due to error when listing contents: %v", listError)
+	}
+
+	if !d.Get("force_destroy").(bool) {
+		deleteErr := fmt.Errorf("Error trying to delete folder without force_destroy set to true")
+		log.Printf("Error! %s : %s\n\n", name, deleteErr)
+		return deleteErr
+	} else {
+		log.Printf("[DEBUG] folder names to delete: %#v", folderName)
+		foldersList, err := config.NewStorageClient(userAgent).Folders.List(bucket).Prefix(folderName[0] + "/").Do()
+		if err != nil {
+			return err
+		}
+		items := foldersList.Items
+		for i := len(items) - 1; i >= 0; i-- {
+			err = transport_tpg.Retry(transport_tpg.RetryOptions{
+				RetryFunc: func() error {
+					err = config.NewStorageClient(userAgent).Folders.Delete(bucket, items[i].Name).Do()
```

Comment:
What if there are objects inside nested folder?

Reply:
we are getting all objects with a top level folder prefix this will list all objects, here in this step only folders will remain so deleting only subfolders and folder here

Reply:
Ohh, I see!
I thought it won't list object that are deeply nested within sub-folders(more than 1 level). 
Got it!